handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java

```diff
@@ -0,0 +1,223 @@
+    private ReadTimeoutTask readTimeoutTask;
+
+    private long readFinishTime;
+    private long writeFinishTime;
+    private ScheduledFuture<?> timeout;
+    private boolean timedOut;
+
+    private volatile boolean removed;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param timeout
+     *        read timeout after last write to the channel
+     * @param unit
+     *        the {@link TimeUnit} of {@code timeout}
+     * @param closeOnTimeout if channel should be closed by this handler after firing {@link ReadTimeoutException}
+     * @param suppressReadAfterTimeout channelRead got after timeout before new write would be ignored if {@code true}.
+     * This address possible situation when we ran into timeout and get response in a very short time
+     * and timeout job executing before channelRead event.
+     * Also this addresses situation when we don't close connection after timeout ({@code closeOnTimeout == false}) and
+     * should skip next read which was received after timeout.
+     * User event {@link ReadSuppressed} will be thrown if case of read suppressing.
+     *
+     */
+    public ReadAfterWriteTimeoutHandler(long timeout, TimeUnit unit, boolean closeOnTimeout,
+                                        boolean suppressReadAfterTimeout) {
+        if (timeout <= 0) {
+            timeoutNanos = 0;
+        } else {
+            timeoutNanos = Math.max(unit.toNanos(timeout), MIN_TIMEOUT_NANOS);
+        }
+        this.closeOnTimeout = closeOnTimeout;
+        this.suppressReadAfterTimeout = suppressReadAfterTimeout;
+    }
+
+    /**
+     * Creates a new instance with {@code closeOnTimeout == true} and {@code ignoreReadAfterTimeout == true}
+     *
+     * @param timeout
+     *        read timeout
+     * @param unit
+     *        the {@link TimeUnit} of {@code timeout}
+     */
+    public ReadAfterWriteTimeoutHandler(long timeout, TimeUnit unit) {
+        this(timeout, unit, true, true);
+    }
+
+    @Override
+    public void handlerAdded(final ChannelHandlerContext ctx) throws Exception {
+        removed = false;
+
+        if (channelOnWriteListener == null) {
+            channelOnWriteListener = new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) throws Exception {
+                    //ensure that only the last write listener in a batch would setup one final timeout task
+                    //or nothing in a case of error
+                    disable();
+                    timedOut = false;
+                    if (future.isSuccess()) {
+                        if (removed) {
+                            return;
+                        }
+                        writeFinishTime = System.nanoTime();
+                        //init readFinishTime to writeFinishTime - readFinishTime > 0 be true if we don't get response
+                        readFinishTime = writeFinishTime - 1;
```

Comment:
Is it necessary to use `nanoTime()` for `readFinishTime` and `writeFinishTime`? It seems like the timeout detection is more based off of if we have received a read since the last time we wrote ... can this be done with a boolean `sawRead` that is set in `channelRead` (maybe based upon something like `timeoutShouldBeCancelled`)  and reset in `channelWrite`?  This handler could be extended and then only reset the `sawRead` for example if the `last message` is detected.